pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeSegmentDataManager.java

```diff
@@ -959,7 +959,13 @@ AtomicBoolean getAcquiredConsumerSemaphore() {
 
   @VisibleForTesting
   SegmentBuildDescriptor buildSegmentInternal(boolean forCommit) {
-    closeStreamConsumers();
+    // for partial upsert tables, do not release _partitionGroupConsumerSemaphore proactively and rely on onDestroy()
```

Comment:
nit: looks like `offload()` closes the consumers

but anyway it looks right to me that the semaphore should be released in the end, when the segment gets replaced